hal/src/boron/network/sara_ncp_client.cpp

```diff
@@ -498,53 +498,63 @@ int SaraNcpClient::getSignalQuality(CellularSignalQuality* qual) {
-        auto resp = parser_.sendCommand("AT+CESQ");
-        int r = CHECK_PARSER(resp.scanf("+CESQ: %d,%d,%d,%d,%d,%d", &rxlev, &rxqual,
-                &rscp, &ecn0, &rsrq, &rsrp));
-        CHECK_TRUE(r == 6, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
-        r = CHECK_PARSER(resp.readResult());
-        CHECK_TRUE(r == AtResponse::OK, SYSTEM_ERROR_AT_NOT_OK);
+        int rsrp;
+        int rsrq_n;
+        unsigned long rsrq_f;
 
-        switch (qual->strengthUnits()) {
-            case CellularStrengthUnits::RXLEV: {
-                qual->strength(rxlev);
-                break;
-            }
-            case CellularStrengthUnits::RSCP: {
-                qual->strength(rscp);
-                break;
-            }
-            case CellularStrengthUnits::RSRP: {
-                qual->strength(rsrp);
-                break;
+        // Set UCGED to mode 5 for RSRP/RSRQ values on R410M
+        CHECK_PARSER_OK(parser_.execCommand("AT+UCGED=5"));
+
+        // Default to invalid value because UCGED command does not explicitly give an ivalid value
+        qual->strength(255);
+        qual->quality(255);
+
+        auto resp = parser_.sendCommand("AT+UCGED?");
+        char buf[1024] = {};     // keep this string long in case there are multiple tower infos in the string
+        while (resp.hasNextLine()) {
+            CHECK_PARSER(resp.readLine(buf, sizeof(buf)));
+            int r = 0;
+            if (strstr(buf, "+RSRP:")) {    // check for RSRP response
+                r = sscanf(buf, "\r\n+RSRP: %*d,%*d,\"%d.%*u\"",&rsrp);
+                CHECK_TRUE(r == 1, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
             }
-            default: {
-                // Do nothing
-                break;
+            if (strstr(buf, "+RSRQ:")) {
+                r = sscanf(buf, "\r\n+RSRQ: %*d,%*d,\"%d.%lu\"",&rsrq_n, &rsrq_f);
+                CHECK_TRUE(r == 2, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
             }
+            buf[0] = '\0';
         }
 
-        switch (qual->qualityUnits()) {
-            case CellularQualityUnits::RXQUAL: {
-                qual->quality(rxqual);
-                break;
-            }
-            case CellularQualityUnits::ECN0: {
-                qual->quality(ecn0);
-                break;
-            }
-            case CellularQualityUnits::RSRQ: {
-                qual->quality(rsrq);
-                break;
-            }
-            default: {
-                // Do nothing
-                break;
-            }
+        if (rsrp < -140) {
+            qual->strength(0);
         }
-    } else {
+        else if (rsrp >= -44) {
+            qual->strength(97);
+        }
+        else {
+            qual->strength(rsrp + 140);
+        }
+
+        int rsrq_mul_100 = rsrq_n * 100 - rsrq_f;
+        if (rsrq_mul_100 < min_rsrq_mul_by_100) {
+            qual->quality(0);
+        }
+        else if (rsrq_mul_100 >= max_rsrq_mul_by_100) {
+            qual->quality(34);
+        }
```

Comment:
Something about bounding the RSRQ value to 0 and 34 is bugging me, because we're supposed to be trusting UCGED more than CSQ and CESQ (which we know can often return 0, 1 and 34 when they shouldn't).  If UCGED happens to return a value outside of that which we expect, how about we return an error... and not a valid value that might be wrong?  If it returns a value outside of this range, it might be due to some other reason like noise or a parsing issue.

The same logic applies to RSRP above with 0 and 97.

We should take a look at Gen 2 with this in mind as well.

Reply:
@technobly 
The problem with UCGED command is that there is no good way to find the value that we **don't expect**. As per the manual and 3GPP specs, everything -inf to +inf can be mapped from 0-34. So we just don't know which values to not expect. Does that sound about right? Please see this pic. Shall we define a range of values to _expect_?

![image](https://user-images.githubusercontent.com/48732259/75723369-c69eb000-5ca1-11ea-80c3-ce87adcad7f5.png)

Reply:
According to my understanding - 
UCGED command just gives the right values when it can read the signal strength. 
When it cannot read the signal strength, it just ERRORs out using an `ERROR` or `CME ERROR`. If it didn't error, then we trust the value that it gives us.